src/esm_runscripts/filelists.py

```diff
@@ -823,9 +826,94 @@ def log_used_files(config):
+
+    logger.debug("\n::: Computing file checksums in ``{target}``")
+    jobtype = config["general"].get("jobtype", "unknown")
+    filetypes = config["general"]["relevant_filetypes"]
+    expid = config["general"]["expid"]
+    it_coupled_model_name = config["general"]["iterative_coupled_model"]
+    datestamp = config["general"]["run_datestamp"]
+    thisrun_log_dir = config["general"]["thisrun_log_dir"]
+    flist_file_yaml = (
+        f"{thisrun_log_dir}/{expid}_{it_coupled_model_name}{jobtype}_filelist_{datestamp}.yaml"
+    )
+    all_files = {}
+
+    checksums = _compute_checksums_for_dir(config, target)
+    files_not_handled_by_filelists = copy.deepcopy(checksums)
+
+    for component in config["general"]["valid_model_names"] + ["general"]:
+        component_files = {}
+        for filetype in filetypes:
+            component_config = config[component]
+
+            for f in component_config.get(f"{filetype}_sources", []):
+
+                target_file = component_config[f"{filetype}_targets"][f]
+                p_target_file = str(pathlib.Path(target_file).absolute())
+
+                checksum = checksums.get(p_target_file, None)
+                if checksum:
+                    del files_not_handled_by_filelists[p_target_file]
+
+                component_files[f] = {
+                    "source": component_config[f"{filetype}_sources"][f],
+                    "intermediate": component_config[f"{filetype}_intermediate"][f],
+                    "target": target_file,
+                    "kind": filetype,
+                    "checksum": checksum,
+                }
+
+                logger.debug(f"::: logging file category: {filetype}")
+                logger.debug(f"- source: {component_files[f]['source']}")
+                logger.debug(f"- target: {component_files[f]['target']}")
+                helpers.print_datetime(config)
+
+        all_files[component] = component_files
+
+    all_files["not_handled_by_filelists"] = {}
+    for f, checksum in files_not_handled_by_filelists.items():
+        all_files["not_handled_by_filelists"][os.path.basename(f)] = {
+            "source": "unknown",
+            "intermediate": "unknown",
+            "target": f,
+            "kind": "not_handled_by_filelists",
+            "checksum": checksum,
+        }
+
+    esm_parser.yaml_dump(all_files, flist_file_yaml)
+
     return config
 
 
+def _compute_checksums_for_dir(config, target):
+
+    if target == "work":
+        dir_path = pathlib.Path(config["general"]["thisrun_work_dir"])
+        file_paths = [str(file.absolute()) for file in dir_path.rglob('*') if file.is_file()]
+    else:
+        logger.error(
+            f"Checksums of files in ``{target}`` directory types are not yet "
+            "supported. Only files in the ``work`` directory currently supported. "
+        )
+        exit(1)
+
+    checksums = {}
+    for f in file_paths:
+        checksums[f] = hashlib.md5(open(f, "rb").read()).hexdigest()
```

Comment:
This is what's slowing it down. I'd think that we can parallelise this for loop, but I am not sure how that will work for gathering in the results. (AFAIK you can't parallel-ly write information into a dictionary.)

Reply:
Indeed, that's the loop that needs to be parallelize. With dask futures, it's pretty straight forward: you make a futures list that contains the md5 processes, each of this processes returns the checksum, then you loop over the futures using the `.result` method (that's your gathering) and then you have the list you need to feed into your dictionary. Once we have the infrastructure for initiating the dask cluster you can take care of this parallelization if you want to learn about it :)